### ISSUE TITLE
Exec Format Error When Deploying FastAPI Example on ARM Architecture

### ISSUE THREAD

**Issue by angelcabo** [awslabs/aws-lambda-web-adapter#311]
I ran into an 'exec format error' while trying to deploy the FastAPI example. I'm on an M1 Mac and attempting to build and run ARM architecture images in lambda. The problem seems to be a potential architecture compatibility issue with the published docker images.

The example Dockerfile is set to copy from the image hosted at [https://gallery.ecr.aws/awsguru/aws-lambda-adapter](https://gallery.ecr.aws/awsguru/aws-lambda-adapter). This repo appears to support multi-arch images, but the image tagged `aarch64` seems to be based on `linux/amd64`.

## Steps to Reproduce
Update the template.yaml with the following to instruct sam to build arm images:
```
Resources:
  FastAPIFunction:
    Type: AWS::Serverless::Function
    Properties:
      Architectures:
        - arm64
```

```
$ sam build --no-cached
Building codeuri: ../aws-lambda-web-adapter/examples/fastapi runtime: None metadata: {'Dockerfile': 'Dockerfile', 'DockerContext': '../aws-lambda-web-adapter/examples/fastapi/app', 'DockerTag': 'python3.8-v1'} architecture: arm64 functions: FastAPIFunction
Building image for FastAPIFunction function
Setting DockerBuildArgs: {} for FastAPIFunction function
Failed building function FastAPIFunction

Build Failed
Error: invalid from flag value public.ecr.aws/awsguru/aws-lambda-adapter:0.7.1-aarch64: image with reference public.ecr.aws/awsguru/aws-lambda-adapter:0.7.
1-aarch64 was found but does not match the specified platform: wanted linux/arm64, actual: linux/amd64
```

<img width="1052" alt="Screenshot 2023-11-18 at 1 32 36 PM" src="https://github.com/awslabs/aws-lambda-web-adapter/assets/478246/7125fc35-37c3-43ab-ad67-d6269ab4ccb3">
<img width="2071" alt="Screenshot 2023-11-18 at 1 32 19 PM" src="https://github.com/awslabs/aws-lambda-web-adapter/assets/478246/26de166e-e542-4b9d-b399-f79799e89a91">

## Workaround

I downloaded the lambda layer and unzipped the extension from there, which seems to contain a 64-bit executable that can be copied to the container directly instead of using the copy from image command in the README.

**Comment by bnusunny** [comment]
@angelcabo Thanks for raising this issue. Could you share your docker and sam cli versions? I'm also using M1 Macbook. I couldn't reproduce this build error here. 

This aarch64 image does seems have the wrong platform. But the binary inside the image is correct.

**Comment by angelcabo** [comment]
```
$ sam --version
SAM CLI, version 1.103.0

$ docker --version
Docker version 24.0.6, build ed223bc
```

I tried running the `docker build --pull` command first and this is what seems to lead to the "exec format error" in the AWS lambda environment for me. It gets around the SAM build issue though because the pull command without specifying a platform pulls the linux/amd64 image with **the arm binary** inside of it. However, the SAM CLI tool builds an x86 image by default (unless you update the architectures in template.yaml). Even though the SAM CLI sets the correct platform target when using the docker client internally to build the image, if an existing lambda web adapter image has already been pulled locally with the same platform metadata, you end up with a SAM built x86 image containing an aarch64 binary inside of it.

My original issue must have been due to at some point pulling the adapter image previously using one architecture and then trying to build with SAM CLI using another, and the SAM CLI wasn't able to ensure the correct image was used for the build. I don't know if this is due to both architecture images for the web adapter being tagged linux/amd64 or if this is just a edgecase to watch out for.
